src/wasm/wasm-stack.cpp

```diff
@@ -2568,25 +2568,44 @@ void BinaryInstWriter::mapLocalsAndEmitHeader() {
   for (Index i = 0; i < func->getNumParams(); i++) {
     mappedLocals[std::make_pair(i, 0)] = i;
   }
+
   // Normally we map all locals of the same type into a range of adjacent
   // addresses, which is more compact. However, if we need to keep DWARF valid,
   // do not do any reordering at all - instead, do a trivial mapping that
   // keeps everything unmoved.
+  //
+  // Unless we have run DWARF-invalidating passes, all locals added during the
+  // process that are not in DWARF info (tuple locals, tuple scratch locals,
+  // locals to resolve stacky format, ..) have been all tacked on to the
+  // existing locals and happen at the end, so as long as we print the local
+  // types in order, we don't invalidate original local DWARF info here.
   if (DWARF) {
-    FindAll<TupleExtract> extracts(func->body);
-    if (!extracts.list.empty()) {
-      Fatal() << "DWARF + multivalue is not yet complete";
+    Index mappedIndex = func->getVarIndexBase();
+    for (Index i = func->getVarIndexBase(); i < func->getNumLocals(); i++) {
+      for (Index j = 0; j < func->getLocalType(i).size(); j++) {
+        mappedLocals[std::make_pair(i, j)] = mappedIndex + j;
+      }
+      mappedIndex += func->getLocalType(i).size();
```

Comment:
Maybe save `func->getLocalType(i).size()` which is used twice to a local `size`?

Reply:
Done